cpp/src/arrow/compute/exec/spilling_join.cc

```diff
@@ -0,0 +1,351 @@
+                        partition_idx_++;
+                        return BuildNextBloomFilter(thread_index);
+                    });
+
+                task_group_probe_[i] = callbacks_.register_task_group(
+                    [this](size_t thread_index, int64_t task_id)
+                    {
+                        return ProbeAndFilterBatch(thread_index, std::move(probe_side_partition_[task_id]));
+                    },
+                    [this](size_t thread_index)
+                    {
+                        return OnProbingFinished(thread_index);
+                    });
+            }
+            return Status::OK();
+        }
+
+        Status SpillingHashJoin::CheckSpilling(size_t thread_index, ExecBatch &batch)
+        {
+            size_t size_of_batch = static_cast<size_t>(batch.TotalBufferSize());
+            size_t max_batch_size = arrow::util::AtomicMax(max_batch_size_, size_of_batch);
+
+            // Spilling algorithm proven to not use more than
+            // (SpillThreshold + NumThreads * BatchSize) memory.
+            // Thus we want to spill when (SpillThreshold + NumThreads * BatchSize) = k * MaxMemory
+            // with some fuzz factor k (which is 0.8 here because that's what I decided). 
+            // Thus SpillThreshold = k * MaxMemory - NumThreads * BatchSize. 
+            constexpr float kFuzzFactor = 0.8f;
+            size_t max_memory = static_cast<size_t>(kFuzzFactor * ctx_->options().max_memory_bytes);
+            size_t spill_threshold =
+                static_cast<size_t>(
+                    std::max(
+                        static_cast<ssize_t>(kFuzzFactor * max_memory - num_threads_ * max_batch_size),
+                        static_cast<ssize_t>(0)));
+            size_t bytes_allocated = static_cast<size_t>(ctx_->memory_pool()->bytes_allocated());
+            size_t bytes_inflight = ctx_->GetCurrentTempFileIO();
+
+            size_t backpressure_threshold = spill_threshold / 2;
+            if(bytes_allocated > backpressure_threshold)
+                callbacks_.pause_probe_side(backpressure_counter_.fetch_add(1));
+            if((bytes_allocated - bytes_inflight) > spill_threshold)
+                RETURN_NOT_OK(AdvanceSpillCursor(thread_index));
+            return Status::OK();
+        }
+
+        Status SpillingHashJoin::AdvanceSpillCursor(size_t thread_index)
+        {
+            bool expected = false;
+            if(!spilling_.load() && spilling_.compare_exchange_strong(expected, true))
+                return callbacks_.start_spilling(thread_index);
+
+            ARROW_ASSIGN_OR_RAISE(bool probe_advanced, probe_accumulator_.AdvanceSpillCursor());
+            if(probe_advanced) return Status::OK();
+
+            ARROW_ASSIGN_OR_RAISE(bool build_advanced, build_accumulator_.AdvanceSpillCursor());
+            if(build_advanced) return Status::OK();
+
+            ARROW_ASSIGN_OR_RAISE(bool probe_hash_advanced, probe_accumulator_.AdvanceHashCursor());
```

Comment:
This looks acceptable but technically there is no need to keep separate files for hashes of probe side batches. Separate files with hashes are only needed for building Bloom filters on build side for spilled partitions. It should be faster to recompute hash after reading data back from disk then writing all hashes to disk and reading back for other cases.

Reply:
I've removed spilling the hashes entirely for now, and will reimplement it when we address the Bloom filter thing.